### PR TITLE
Create new projects for System.Runtime.Analyzers

### PR DESCRIPTION
**Pull request by srivatsn**
This is the start of work to split up the FxCop analyzers into multiple packages that can ship with the APIs that they address.

Move rule CA1001 to this project. Rewrite the codefixer for it to be based on DocumentEditor and hence language agnostic.

Fix for CA1001 shouldn't generate a dispose method if there is already one in the type. Instead modify that method to make it the interface implementation. Adding some tests.

### CODE REVIEW

File: src/Diagnostics/FxCop/CSharp/CSharpFxCopRulesDiagnosticAnalyzers.csproj [hunk 1]
```diff
@@ -107,4 +106,4 @@
     <Import Project="..\..\..\..\build\VSL.Imports.Closed.targets" />
     <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
   </ImportGroup>
-</Project>
+</Project>
```
Comment (dpoeschl) [inline src/Diagnostics/FxCop/CSharp/CSharpFxCopRulesDiagnosticAnalyzers.csproj]: I think @Pilchie might have said something in another review about newlines at the end of proj files, but I can't remember why they mattered.

Comment (srivatsn) [inline src/Diagnostics/FxCop/CSharp/CSharpFxCopRulesDiagnosticAnalyzers.csproj]: Ah yes. Fixed them.

Comment (Pilchie) [inline src/Diagnostics/FxCop/CSharp/CSharpFxCopRulesDiagnosticAnalyzers.csproj]: It's not the end of the world, but many Unix utils don't like files that don't end in a line ending.
--
Kevin Pilch-Bisson
    http://twitter.com/Pilchie
________________________________

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Design/TypesThatOwnDisposableFieldsShouldBeDisposable.cs [hunk 2]
```diff
@@ -1,34 +1,32 @@
...
 using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
 using Microsoft.CodeAnalysis;
 using Microsoft.CodeAnalysis.Diagnostics;
-using Microsoft.CodeAnalysis.FxCopAnalyzers.Utilities;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.FxCopAnalyzers.Design
+namespace System.Runtime.Analyzers
```
Comment (tmeschter) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Design/TypesThatOwnDisposableFieldsShouldBeDisposable.cs]: Do we need to check with anyone before we start adding namespaces under `System`?

**Comment by tmeschter** [comment]
:+1:

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Design/TypesThatOwnDisposableFieldsShouldBeDisposable.cs (hunk 2 above)
Comment (srivatsn) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/Design/TypesThatOwnDisposableFieldsShouldBeDisposable.cs]: Not that I know of.

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/CodeFixProviderBase.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+
+namespace System.Runtime.Analyzers
+{
+    public abstract class CodeFixProviderBase : CodeFixProvider
```
Comment (mavasani) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/CodeFixProviderBase.cs]: Is CodeFixProviderBase used anywhere in this project?

Comment (srivatsn) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/CodeFixProviderBase.cs]: No I've deleted this on my machine. That'll be part of the next change.

File: src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/DiagnosticExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+
+namespace System.Runtime.Analyzers
+{
+    internal static class DiagnosticExtensions
```
Comment (mavasani) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/DiagnosticExtensions.cs]: All these seem like good helpers which are useful for all kind of analyzers. Probably someday we can make these public in CodeAnalysis layer and avoid duplication?

Comment (srivatsn) [inline src/Diagnostics/FxCop/System.Runtime.Analyzers/Core/DiagnosticExtensions.cs]: Yes I'll collect a list of these as I make move more of this stuff around.

**Comment by mavasani** [comment]
:+1:

**Pull request merged**